snap/errors.go

```diff
@@ -41,10 +45,43 @@ func (e NotInstalledError) Error() string {
 	return fmt.Sprintf("revision %s of snap %q is not installed", e.Rev, e.Snap)
 }
 
+// NotSnapError is returned if a operation expects a snap file or snap dir
+// but no valid input is provided.
 type NotSnapError struct {
 	Path string
+
+	Err error
 }
 
 func (e NotSnapError) Error() string {
-	return fmt.Sprintf("%q is not a snap or snapdir", e.Path)
+	return fmt.Sprintf("cannot process snap or snapdir: %v", e.Err)
+}
+
+func buildNotSnapErrorContext(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	stat, err := f.Stat()
+	if err != nil {
+		return err
+	}
+	if stat.IsDir() {
+		if _, err := f.Readdir(1); err == io.EOF {
+			return fmt.Errorf("directory %q is empty", path)
+		}
+		return fmt.Errorf("directory %q is invalid", path)
+	}
+
+	var header [10]byte
+	if _, err := f.Read(header[:]); err != nil {
+		return err
+	}
+	return fmt.Errorf("file %q is invalid (header %v)", path, header)
+}
```

Comment:
it's strange for this logic to live here, it seems it should live in snapfile.go itself, that one has a sense of directory vs not

Reply:
Yeah, it's a bit awkward, my thinking was mostly that this error is about NotSnapError so this does not know any details about why it's not a snap just gives an arbitrary number of header bytes or basic stat data (file missing, directory).

An alternatively idea would be to change `snapfile.go` and detect this inside there and set the NoSnapError msg there?

Reply:
my preference is definitely to have this logic in snapfile.go, also because over time we might want to move away from it being generic and actually doing some more consistent checks?

Reply:
Thanks! I moved it.